substrate/frame/assets/src/functions.rs

```diff
@@ -361,7 +386,8 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			T::Currency::unreserve(&who, deposit);
 		}
 
-		if let Remove = Self::dead_account(&who, &mut details, &account.reason, false) {
+		if let Remove = Self::dead_account(id.clone(), &who, &mut details, &account.reason, false)?
```

Comment:
`do_refund` is not strictly a call implementation. It is also used in traits in the implementation of `fungibles::Refund`.

The trait `fungibles::Refund` doesn't specifies that the storage must be rollbacked in case of error.


Somebody can write:
```
// Ignore because we don't need to enforce the success.
let _ignore = fungibles::Refund::refund(asset_id, some_account);
```
And now storage is inconsistent.

Reply:
see https://github.com/paritytech/polkadot-sdk/issues/6342

Reply:
I understand you're proposing these `T::Currency::unreserve` calls should be wrapped inside a `transactional` macro? Or does the entire call need to be wrapped?

Reply:
I guess wrapping from the `unreserve` call until `dead_account` or the end of the block into one transactional is more efficient.
Because the early checks can return without starting the transactional.

Reply:
Done in [`d2e2d18`](https://github.com/paritytech/polkadot-sdk/pull/4530/commits/d2e2d18a96ffc0af8ad041dce1a0d9a22a1cd334)